### ISSUE TITLE
batch insert breaks after 5k lines with ca 65 columns

### ISSUE THREAD

**Issue by ArdentZeal** [tgriesser/knex#757]
Hello there,

I tried to write an Excel import for my application. After importing smaller excelfiles (1000ish lines), I tried a big one with ca 20000. The excel import stopped working.

I debugged it down to the parameterized queries, which seem to not be able to handle all the different parameters and is ULTRA slow at that point too.

To workaround that, I used the toString() method and applied the query then with knex.raw.
The problem with this approach is that the 5k line limit is still valid. Even the toString()-Method seems to be at some kind of end with too much data.

So my second workaroung is to bundle my data in smaller packets (1000) and to toString() that (to be fast again), which works but is really really ugly.

```
 var sqlquery = '';

                var batchSize = 1000;
                var newListEntriesLength = newListEntries.length;
                // build sql string in batchSize packs
                for(var i = 0; (i * batchSize) < newListEntriesLength; i++) {
                    var slicedNewEntries = newListEntries.splice(0, batchSize);

                    if(slicedNewEntries.length > 0) {
                        sqlquery += _strucd.knex(tableName).insert(slicedNewEntries).toString() + '; ';
                    } 
                }

                // Add leftovers
                if(newListEntries.length > 0) {
                    sqlquery += _strucd.knex(tableName).insert(newListEntries).toString() + ';';
                }
                _strucd.knex.raw(sqlquery)
                    .then(function(results) {
                        return callback(null, results);
                    }, function(err) {
                        return callback(err);
                    });
```

1. Am I missing something? 
2. Is it really that complicated to do fast batch insert with several 1000 lines?
3. Is there a better way to build an insert query, without doing everything it manually?

Kind Regards,
Matthias

**Comment by tgriesser** [comment]
I'm going to dig into this a bit and see if I can improve things to make this possible

**Comment by ArdentZeal** [comment]
Ah Im using Postgres underneath, if that matters at all.

**Comment by tgriesser** [comment]
At the very least I'll make it so there's a (configurable) batch limit so the ugly code is pushed into the library and not on you to produce

**Comment by catmsred** [comment]
Just checking in to see if there's been any work on this?  I'm also working with big imports (~16000 lines) on MariaDB, and I'm having similar problem.  If I let the query run long enough, then I eventually get the following error:

```
[project]/node_modules/mariasql/lib/Client.js:367
      this._client.query(this._curResults._query, this._curResults._useArray);
                   ^
Error: Not ready to query
    at Error (native)
    at Client._processQueries ([project]/node_modules/mariasql/lib/Client.js:367:20)
    at [project]/node_modules/mariasql/lib/Client.js:263:38
    at process._tickCallback (node.js:355:11)
```

Which makes it look like the problem might be with MariaDB, except that the OP has a similar result using Postres.

I also found that if you're using Bluebird promises, you can use a shorter hack using the Bluebird Promise.map concurrency option, as follows:

```
    Promise.map(data, function(row) {
      return knex.table(tablename)
        .insert(row)
        .then(function() {
          //console.log('row inserted');
        });
    }, {concurrency: 1})
    .then( function() {
      console.log('done');
    })
    .catch( function(e) {
      console.error(e);
    });
```

This means that only one short db connection is open at any given time.  More than one might be okay, but usually there's a max limit to that, so I decided to start small.

**Comment by tgriesser** [comment]
Yeah so I'm thinking there should be an API like:

```js
knex.batchInsert(values, limit).into(table)
```

Which would operate by inserting values in batches of limit (where limit would default to 1), and it would automatically wrap everything in a transaction.

Does this sound reasonable?

**Comment by ArdentZeal** [comment]
sounds fine by me.

**Comment by catmsred** [comment]
Yeah, that would be great.  Thanks!

**Comment by tgriesser** [comment]
Can you try the `batch-insert` branch I just pushed and see if it's intutive:

API:

```js
knex.batchInsert(table, values, [limit: default 1]).then(...
```

**Comment by ArdentZeal** [comment]
Hey,

I tried your branch with a batchSize of 1000. The interface works, the data is inserted, BUT my handwritten solution takes like 4 seconds, your solution around 21s, so I prefer the crappy, fast code at the moment :)

```
_strucd.knex()
                    .batchInsert(tableName, newListEntries, batchSize)
                    .then(function(results) {
                        return callback(null);
                    }, function(err) {
                        return callback(err);
                    }); 
```

**Comment by ArdentZeal** [comment]
I am using the batch-insert feature at the moment, any chance it will be merged into a new release so I can get all the bugfixes again?

**Comment by catmsred** [comment]
I haven't had a chance to test this extensively (I got moved to a more urgent project temporarily), but I'd like to second the request to have this feature merged into the new release.  Thanks!

**Comment by viig99** [comment]
+1

**Comment by mkko** [comment]
+1

**Comment by ArdentZeal** [comment]
can we get proper batchInsert support, with transactions, with returning into the next release? This issue is open since  Mar 31.

Kind Regards,
ArdentZeal

**Comment by rhys-vdw** [comment]
Hey everyone. Obviously you all want `batchInsert`. I'm going to try to do a release soon, and it's not going to make it in because I have no pull request to review, and the branch is stale and targets the old `lib/` build folder.

However once I've learned how to do releases, it should be simple enough for me to do another release soon. I'll just merge and release when the PR is ready.

Will you be able to create a new PR @ArdentZeal?

**Comment by ArdentZeal** [comment]
I did not wrote the original logic for the batch insert.
@tgriesser did that. It never got fully completed sadly.

I am not very well versed in the knexjs internals (read: never looked at the internal code for more than 1 minute) and I am really limited for time at the moment, so I cant promise a PR.

Kind Regards

**Comment by wubzz** [comment]
I've made a pull request for this: #1182

**Comment by elhigu** [comment]
Fixed by #1182
